Makefile

```diff
@@ -20,27 +20,58 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
-# Service
-NAME ?= cray-sls
-VERSION ?= $(shell cat .version)
+# import config.
+# You can change the default config with `make cnf="config_special.env" build`
+cnf ?= config.env
+include $(cnf)
+export $(shell sed 's/=.*//' $(cnf))
 
+ifndef NAME
+$(error NAME is not set.  Please review and copy config.env.default to config.env and try again)
+endif
+
+# HELP
+# This will output the help for each task
+# thanks to https://marmelab.com/blog/2016/02/29/auto-documented-makefile.html
+.PHONY: help build-image touch.build-image
+
+help: ## This help.
+	@awk 'BEGIN {FS = ":.*?## "} /^[a-zA-Z_-]+:.*?## / {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}' $(MAKEFILE_LIST)
+
+.DEFAULT_GOAL := help
+
+# Set the basic docker build command
+
+DOCKER_BUILD_CMD=docker build --build-arg REGISTRY_HOST=${REGISTRY_HOST} ${DOCKER_ARGS}
 
 all : image unittest ct snyk ct_image
 
-image:
-	docker build ${NO_CACHE} --pull ${DOCKER_ARGS} --tag '${NAME}:${VERSION}' .
+image: touch.build-image .build-image ## Build the service image
+
+.build-image: Dockerfile 
+	${DOCKER_BUILD_CMD} --no-cache --pull --tag '${NAME}:${VERSION}' .
+	${DOCKER_BUILD_CMD} --target builder --tag '${NAME}-build:${VERSION}' .
+	@touch .build-image
 
-unittest:
+test-image: image  ## Build the image for running tests
+	${DOCKER_BUILD_CMD} --tag '${NAME}-test:${VERSION}' -f Dockerfile.testing
+	
+
+unittest: image ## Run the unittests^Wintegration tests
 	./runUnitTest.sh
 
-snyk:
+snyk: ## Run Snyk
 	./runSnyk.sh
 
-ct:
+ct: ## Run Conformance Tests
 	./runCT.sh
 
 ct_image:
 	docker build --no-cache -f test/ct/Dockerfile test/ct/ --tag hms-sls-hmth-test:${VERSION}
 
 sls-loader-helper:
-	docker build --no-cache -f test/ct/Dockerfile.load-sls.Dockerfile test/ct/ --tag hms-sls-loader:${VERSION}
\ No newline at end of file
+	docker build --no-cache -f test/ct/Dockerfile.load-sls.Dockerfile test/ct/ --tag hms-sls-loader:${VERSION}
+
+touch.build-image:
+	$(eval timestamp=$(shell docker inspect -f '{{ range $$i, $$e := split .Metadata.LastTagTime "T" }}{{if eq $$i 0}}{{range $$j, $$v := split $$e "-"}}{{$$v}}{{end}}{{else}}{{$$f := printf "%.8s" $$e}}{{range $$j, $$g := split $$f ":"}}{{if lt $$j 2}}{{$$g}}{{else}}.{{$$g}}{{end}}{{end}}{{end}}{{end}}' $(DOCKER_TAG) 2>/dev/null ))
```

Comment:
A comment here on what this is doing would be helpful.

Reply:
You mean this isn't self-documenting? :)

I'll update the PR with a description and/or simplify this.